File: typescript/src/resources/gemini.ts
/** Gemini resources. */

import { TaskHandle } from '../runtime/tasks';
import { Transport } from '../runtime/transport';

export type GeminiModel =
  | 'gemini-3.7-flash'
  | 'gemini-3.6-flash'
  | 'gemini-3.5-flash'
  | 'gemini-3.5-flash-lite'
  | 'gemini-3.1-flash-lite'
  | 'gemini-3.1-pro-preview'
  | 'gemini-3-flash-preview'
  | 'gemini-2.5-pro'
  | 'gemini-2.5-flash'
  | 'gemini-2.5-flash-lite'
  | 'gemini-3.1-flash-image'
  | 'gemini-2.5-flash-image'
  | 'gemini-3-pro-image';

function taskId(result: Record<string, unknown>): string {
  return typeof result.task_id === 'string' ? result.task_id : typeof result.id === 'string' ? result.id : '';
}

export class Gemini {
  constructor(private transport: Transport) {}

  async chatCompletions(opts: {
    model: GeminiModel;
    messages: Array<Record<string, unknown>>;
    stream?: false;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>>;
  async chatCompletions(opts: {
    model: GeminiModel;
    messages: Array<Record<string, unknown>>;
    stream: true;
    [key: string]: unknown;
  }): Promise<AsyncGenerator<Record<string, unknown>>>;
  async chatCompletions(opts: {
    model: GeminiModel;
    messages: Array<Record<string, unknown>>;
    stream?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown> | AsyncGenerator<Record<string, unknown>>> {
    const { stream, ...body } = opts;
    if (stream) return this.stream('/gemini/chat/completions', { ...body, stream: true });
    return this.transport.request('POST', '/gemini/chat/completions', { json: body });
  }

  async generateContent(
    model: GeminiModel,
    contents: Array<Record<string, unknown>>,
    extra: Record<string, unknown> = {}
  ): Promise<Record<string, unknown>> {
    return this.transport.request('POST', `/v1beta/models/${model}:generateContent`, {
      json: { contents, ...extra },
    });
  }

  streamGenerateContent(
    model: GeminiModel,
    contents: Array<Record<string, unknown>>,
    extra: Record<string, unknown> = {}
  ): AsyncGenerator<Record<string, unknown>> {
    return this.stream(`/v1beta/models/${model}:streamGenerateContent`, { contents, ...extra });
  }

  async generateVideo(opts: {
    prompt: string;
    model?: 'omni-flash';
    aspectRatio?: '16:9' | '9:16';
    resolution?: '720p' | '1080p';
    imageUrls?: string[];
    videoUrls?: string[];
    async?: boolean;
    wait?: boolean;
    pollInterval?: number;
    maxWait?: number;
    callbackUrl?: string;
  }): Promise<TaskHandle> {
    const { async: async_, wait, pollInterval, maxWait, callbackUrl, imageUrls, videoUrls, ...rest } = opts;
    const body: Record<string, unknown> = {
      ...rest,
      model: opts.model ?? 'omni-flash',
      aspect_ratio: opts.aspectRatio ?? '16:9',
      resolution: opts.resolution ?? '720p',
      async: async_ ?? true,
    };
    if (imageUrls !== undefined) body.image_urls = imageUrls;
    if (videoUrls !== undefined) body.video_urls = videoUrls;
    if (callbackUrl !== undefined) body.callback_url = callbackUrl;
    const result = (await this.transport.request('POST', '/gemini/videos', { json: body })) as Record<string, unknown>;
    const handle = new TaskHandle(taskId(result), '/gemini/tasks', this.transport, result);
    if (wait) await handle.wait({ pollInterval, maxWait });
    return handle;
  }

  async tasks(opts: { id?: string; ids?: string[]; action?: 'retrieve' | 'retrieve_batch' } = {}): Promise<Record<string, unknown>> {
    const { id, ids, action = 'retrieve' } = opts;
    const body: Record<string, unknown> = { action };
    if (id !== undefined) body.id = id;
    if (ids !== undefined) body.ids = ids;
    return this.transport.request('POST', '/gemini/tasks', { json: body });
  }

  private async *stream(path: string, body: Record<string, unknown>): AsyncGenerator<Record<string, unknown>> {
    for await (const chunk of this.transport.requestStream('POST', path, { json: body })) yield JSON.parse(chunk);
  }
}

File: python/src/acedatacloud/resources/gemini.py
"""Gemini resources."""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Iterator
from typing import Any, Literal

from acedatacloud._runtime.tasks import AsyncTaskHandle, TaskHandle

GeminiModel = Literal[
    "gemini-3.7-flash", "gemini-3.6-flash", "gemini-3.5-flash", "gemini-3.5-flash-lite",
    "gemini-3.1-flash-lite", "gemini-3.1-pro-preview", "gemini-3-flash-preview",
    "gemini-2.5-pro", "gemini-2.5-flash", "gemini-2.5-flash-lite",
    "gemini-3.1-flash-image", "gemini-2.5-flash-image", "gemini-3-pro-image",
]


def _task_id(result: dict[str, Any]) -> str:
    return str(result.get("task_id") or result.get("id") or "")


class Gemini:
    """Synchronous Gemini client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def chat_completions(
        self, *, model: GeminiModel, messages: list[dict[str, Any]], stream: bool = False, **extra: Any
    ) -> dict[str, Any] | Iterator[dict[str, Any]]:
        body = {"model": model, "messages": messages, **extra}
        if stream:
            body["stream"] = True
            return (json.loads(chunk) for chunk in self._transport.request_stream("POST", "/gemini/chat/completions", json=body))
        return self._transport.request("POST", "/gemini/chat/completions", json=body)

    def generate_content(self, *, model: GeminiModel, contents: list[dict[str, Any]], **extra: Any) -> dict[str, Any]:
        return self._transport.request(
            "POST", f"/v1beta/models/{model}:generateContent", json={"contents": contents, **extra}
        )

    def stream_generate_content(
        self, *, model: GeminiModel, contents: list[dict[str, Any]], **extra: Any
    ) -> Iterator[dict[str, Any]]:
        body = {"contents": contents, **extra}
        return (json.loads(chunk) for chunk in self._transport.request_stream(
            "POST", f"/v1beta/models/{model}:streamGenerateContent", json=body
        ))

    def generate_video(
        self, *, prompt: str, model: Literal["omni-flash"] = "omni-flash",
        aspect_ratio: Literal["16:9", "9:16"] = "16:9",
        resolution: Literal["720p", "1080p"] = "720p", image_urls: list[str] | None = None,
        video_urls: list[str] | None = None, async_: bool | None = None, wait: bool = False,
        poll_interval: float = 3.0, max_wait: float = 600.0, callback_url: str | None = None,
    ) -> TaskHandle:
        body: dict[str, Any] = {"prompt": prompt, "model": model, "aspect_ratio": aspect_ratio, "resolution": resolution}
        if image_urls is not None: body["image_urls"] = image_urls
        if video_urls is not None: body["video_urls"] = video_urls
        if callback_url is not None: body["callback_url"] = callback_url
        body["async"] = True if async_ is None else async_
        result = self._transport.request("POST", "/gemini/videos", json=body)
        handle = TaskHandle(_task_id(result), "/gemini/tasks", self._transport, submitted=result)
        if wait: handle.wait(poll_interval=poll_interval, max_wait=max_wait)
        return handle

    def tasks(
        self, *, id: str | None = None, ids: list[str] | None = None,
        action: Literal["retrieve", "retrieve_batch"] = "retrieve",
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"action": action}
        if id is not None: body["id"] = id
        if ids is not None: body["ids"] = ids
        return self._transport.request("POST", "/gemini/tasks", json=body)


class AsyncGemini:
    """Asynchronous Gemini client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def chat_completions(
        self, *, model: GeminiModel, messages: list[dict[str, Any]], stream: bool = False, **extra: Any
    ) -> dict[str, Any] | AsyncIterator[dict[str, Any]]:
        body = {"model": model, "messages": messages, **extra}
        if stream:
            body["stream"] = True
            return self._stream("/gemini/chat/completions", body)
        return await self._transport.request("POST", "/gemini/chat/completions", json=body)

    async def generate_content(self, *, model: GeminiModel, contents: list[dict[str, Any]], **extra: Any) -> dict[str, Any]:
        return await self._transport.request(
            "POST", f"/v1beta/models/{model}:generateContent", json={"contents": contents, **extra}
        )

    async def stream_generate_content(
        self, *, model: GeminiModel, contents: list[dict[str, Any]], **extra: Any
    ) -> AsyncIterator[dict[str, Any]]:
        return self._stream(f"/v1beta/models/{model}:streamGenerateContent", {"contents": contents, **extra})

    async def generate_video(
        self, *, prompt: str, model: Literal["omni-flash"] = "omni-flash",
        aspect_ratio: Literal["16:9", "9:16"] = "16:9",
        resolution: Literal["720p", "1080p"] = "720p", image_urls: list[str] | None = None,
        video_urls: list[str] | None = None, async_: bool | None = None, wait: bool = False,
        poll_interval: float = 3.0, max_wait: float = 600.0, callback_url: str | None = None,
    ) -> AsyncTaskHandle:
        body: dict[str, Any] = {"prompt": prompt, "model": model, "aspect_ratio": aspect_ratio, "resolution": resolution}
        if image_urls is not None: body["image_urls"] = image_urls
        if video_urls is not None: body["video_urls"] = video_urls
        if callback_url is not None: body["callback_url"] = callback_url
        body["async"] = True if async_ is None else async_
        result = await self._transport.request("POST", "/gemini/videos", json=body)
        handle = AsyncTaskHandle(_task_id(result), "/gemini/tasks", self._transport, submitted=result)
        if wait: await handle.wait(poll_interval=poll_interval, max_wait=max_wait)
        return handle

    async def tasks(
        self, *, id: str | None = None, ids: list[str] | None = None,
        action: Literal["retrieve", "retrieve_batch"] = "retrieve",
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"action": action}
        if id is not None: body["id"] = id
        if ids is not None: body["ids"] = ids
        return await self._transport.request("POST", "/gemini/tasks", json=body)

    async def _stream(self, path: str, body: dict[str, Any]) -> AsyncIterator[dict[str, Any]]:
        async for chunk in self._transport.request_stream("POST", path, json=body):
            yield json.loads(chunk)
